plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsReader.java

```diff
@@ -276,41 +272,35 @@ private static Map<Integer, Long> readNdvs(Table icebergTable, long snapshotId,
     }
 
     /**
-     * Iterates over existing statistics files present for parent snapshot chain,  starting at {@code startingSnapshotId} (inclusive).
+     * Returns most recent statistics file for the given {@code snapshotId}
      */
-    public static Iterator<StatisticsFile> walkStatisticsFiles(Table icebergTable, long startingSnapshotId)
+    public static Optional<StatisticsFile> getStatisticsFile(Table icebergTable, long snapshotId)
     {
-        return new AbstractIterator<>()
-        {
-            private final Map<Long, StatisticsFile> statsFileBySnapshot = icebergTable.statisticsFiles().stream()
-                    .collect(toMap(
-                            StatisticsFile::snapshotId,
-                            identity(),
-                            (a, b) -> {
-                                throw new IllegalStateException("Unexpected duplicate statistics files %s, %s".formatted(a, b));
-                            },
-                            HashMap::new));
-
-            private final Iterator<Long> snapshots = walkSnapshots(icebergTable, startingSnapshotId);
-
-            @Override
-            protected StatisticsFile computeNext()
-            {
-                if (statsFileBySnapshot.isEmpty()) {
-                    // Already found all statistics files
-                    return endOfData();
-                }
+        Map<Long, StatisticsFile> statsFileBySnapshot = icebergTable.statisticsFiles().stream()
+                .collect(toMap(
+                        StatisticsFile::snapshotId,
+                        identity(),
+                        (a, b) -> {
+                            throw new IllegalStateException("Unexpected duplicate statistics files %s, %s".formatted(a, b));
+                        },
+                        HashMap::new));
 
-                while (snapshots.hasNext()) {
-                    long snapshotId = snapshots.next();
-                    StatisticsFile statisticsFile = statsFileBySnapshot.remove(snapshotId);
-                    if (statisticsFile != null) {
-                        return statisticsFile;
-                    }
-                }
-                return endOfData();
+        Iterator<Long> snapshots = walkSnapshots(icebergTable, snapshotId);
+        if (statsFileBySnapshot.isEmpty()) {
+            // Already found all statistics files
+            return Optional.empty();
+        }
+        if (statsFileBySnapshot.size() == 1) {
+            return statsFileBySnapshot.values().stream().findAny();
+        }
+        while (snapshots.hasNext()) {
+            long snapshot = snapshots.next();
+            StatisticsFile statisticsFile = statsFileBySnapshot.remove(snapshot);
```

Comment:
Why remove here? The map isn't used if the value is found, so I don't see the point in modifying it.

Seems this is leftover from before, and we can change to the three-argument version of `toMap()` that returns an unmodifiable map.